test/unit/style-spec/feature_filter.test.js

```diff
@@ -71,6 +72,36 @@ test('filter', t => {
     t.end();
 });
 
+test('legacy filter detection', t => {
+    t.test('definitely legacy filters', t => {
+        // Expressions with more than two arguments.
+        t.notOk(isExpressionFilter(["in", "color", "red", "blue"]));
+
+        // Expressions where the second argument is not a string or array.
+        t.notOk(isExpressionFilter(["in", "value", 42]));
+        t.notOk(isExpressionFilter(["in", "value", true]));
+        t.end();
+    });
+
+    t.test('ambiguous value', t => {
+        // Should err on the side of reporting as a legacy filter. Style authors can force filters
+        // by using a literal expression as the first argument.
```

Comment:
This workaround might be worth mentioning in the style specification documentation.